File: omrs/src/main/java/org/apache/atlas/omrs/metadatacollection/properties/typedefs/TypeDef.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.atlas.omrs.metadatacollection.properties.typedefs;

import com.fasterxml.jackson.annotation.*;
import org.apache.atlas.omrs.metadatacollection.properties.instances.*;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.PUBLIC_ONLY;

/**
 * The TypeDef is the base class for objects that store the properties of an open metadata type
 * definition (call ed a TypeDef).
 * <p>
 * The different categories of Typedefs are listed in TypeDefCategory.
 */
@JsonAutoDetect(getterVisibility=PUBLIC_ONLY, setterVisibility=PUBLIC_ONLY, fieldVisibility=NONE)
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown=true)
@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = EntityDef.class, name = "EntityDef"),
        @JsonSubTypes.Type(value = RelationshipDef.class, name = "RelationshipDef"),
        @JsonSubTypes.Type(value = ClassificationDef.class, name = "ClassificationDef"),
})
public class TypeDef extends TypeDefSummary
{
    protected TypeDefLink                        superType                = null;
    protected String                             description              = null;
    protected String                             descriptionGUID          = null;
    protected String                             origin                   = null;
    protected String                             createdBy                = null;
    protected String                             updatedBy                = null;
    protected Date                               createTime               = null;
    protected Date                               updateTime               = null;
    protected Map<String, String>                options                  = null;
    protected ArrayList<ExternalStandardMapping> externalStandardMappings = null;
    protected ArrayList<InstanceStatus>          validInstanceStatusList  = null;
    protected InstanceStatus                     initialStatus            = null;
    protected ArrayList<TypeDefAttribute>        propertiesDefinition     = null;


    /**
     * Default constructor
     */
    public TypeDef()
    {
    }


    /**
     * Minimal constructor is passed the category of the typedef being constructed.
     * The rest of the properties are null.
     *
     * @param category - TypeDefCategory enum
     */
    public TypeDef(TypeDefCategory category)
    {
        super();
        this.category = category;
    }


    /**
     * Typical constructor is passed the properties of the typedef's super class being constructed.
     *
     * @param category    - category of this TypeDef
     * @param guid        - unique id for the TypeDef
     * @param name        - unique name for the TypeDef
     * @param version     - active version number for the TypeDef
     * @param versionName - name for the active version of the TypeDef
     */
    public TypeDef(TypeDefCategory category,
                   String          guid,
                   String          name,
                   long            version,
                   String          versionName)
    {
        super(category, guid, name, version, versionName);
    }


    /**
     * Copy/clone constructor copies the values from the supplied template.
     *
     * @param template TypeDef
     */
    public TypeDef(TypeDef template)
    {
        super(template);

        if (template != null)
        {
            this.superType = template.getSuperType();
            this.description = template.getDescription();
            this.descriptionGUID = template.getDescriptionGUID();
            this.origin = template.getOrigin();
            this.createdBy = template.getCreatedBy();
            this.updatedBy = template.getUpdatedBy();
            this.createTime = template.getCreateTime();
            this.updateTime = template.getUpdateTime();
            this.options = template.getOptions();
            this.setExternalStandardMappings(template.getExternalStandardMappings());
            this.setValidInstanceStatusList(template.getValidInstanceStatusList());
            this.setPropertiesDefinition(template.getPropertiesDefinition());
        }
    }


    /**
     * Return the super type for the TypeDef (or null if top-level)
     *
     * @return TypeDefLink for the super type
     */
    public TypeDefLink getSuperType()
    {
        return superType;
    }


    /**
     * Set up supertype for the TypeDef.  Only single inheritance is supported.  Use null if this type
     * is top-level.
     *
     * @param superType TypeDefLink for the super type
     */
    public void setSuperType(TypeDefLink superType) { this.superType = superType; }


    /**
     * Return the description of this TypeDef.
     *
     * @return String description
     */
    public String getDescription()
    {
        return description;
    }


    /**
     * Set up the description of this TypeDef.
     *
     * @param description String
     */
    public void setDescription(String description)
    {
        this.description = description;
    }


    /**
     * Return the unique identifier (guid) of the glossary term that describes this TypeDef.
     *
     * @return String guid
     */
    public String getDescriptionGUID()
    {
        return descriptionGUID;
    }


    /**
     * Set up the unique identifier (guid) of the glossary term that describes this TypeDef.
     *
     * @param descriptionGUID - String guid
     */
    public void setDescriptionGUID(String descriptionGUID)
    {
        this.descriptionGUID = descriptionGUID;
    }


    /**
     * Return the unique identifier for metadata collection Id where this TypeDef came from.
     *
     * @return String guid
     */
    public String getOrigin()
    {
        return origin;
    }


    /**
     * Set up the unique identifier for metadata collection Id where this TypeDef came from.
     *
     * @param origin - String guid
     */
    public void setOrigin(String origin)
    {
        this.origin = origin;
    }


    /**
     * Return the user name of the person that created this TypeDef.
     *
     * @return String name
     */
    public String getCreatedBy()
    {
        return createdBy;
    }


    /**
     * Set up the user name of the person that created this TypeDef.
     *
     * @param createdBy String name
     */
    public void setCreatedBy(String createdBy)
    {
        this.createdBy = createdBy;
    }


    /**
     * Return the user name of the person that last updated this TypeDef.
     *
     * @return String name
     */
    public String getUpdatedBy()
    {
        return updatedBy;
    }


    /**
     * Set up the user name of the person that last updated this TypeDef.
     *
     * @param updatedBy String name
     */
    public void setUpdatedBy(String updatedBy)
    {
        this.updatedBy = updatedBy;
    }


    /**
     * Return the date/time that this TypeDef was created.
     *
     * @return Date
     */
    public Date getCreateTime()
    {
        return createTime;
    }


    /**
     * Set up the date/time that this TypeDef was created.
     *
     * @param createTime Date
     */
    public void setCreateTime(Date createTime)
    {
        this.createTime = createTime;
    }


    /**
     * Return the date/time that this TypeDef was last updated.
     *
     * @return Date
     */
    public Date getUpdateTime()
    {
        return updateTime;
    }


    /**
     * Set up the date/time that this TypeDef was last updated.
     *
     * @param updateTime Date
     */
    public void setUpdateTime(Date updateTime)
    {
        this.updateTime = updateTime;
    }


    /**
     * Return the options for this TypeDef. These are private properties used by the processors of this TypeDef
     * and ignored by the OMRS.
     *
     * @return Map from String to String
     */
    public Map<String, String> getOptions()
    {
        return options;
    }


    /**
     * Set up the options for this TypeDef.  These are private properties used by the processors of this TypeDef
     * and ignored by the OMRS.
     *
     * @param options - Map from String to String
     */
    public void setOptions(Map<String, String> options)
    {
        this.options = options;
    }


    /**
     * Return the list of mappings to external standards.
     *
     * @return ExternalStandardMappings list
     */
    public List<ExternalStandardMapping> getExternalStandardMappings()
    {
        if (externalStandardMappings == null)
        {
            return null;
        }
        else
        {
            return new ArrayList<>(externalStandardMappings);
        }
    }


    /**
     * Set up the list of mappings to external standards.
     *
     * @param externalStandardMappings - ExternalStandardMappings list
     */
    public void setExternalStandardMappings(List<ExternalStandardMapping> externalStandardMappings)
    {
        if (externalStandardMappings == null)
        {
            this.externalStandardMappings = null;
        }
        else
        {
            this.externalStandardMappings = new ArrayList<>(externalStandardMappings);
        }
    }


    /**
     * Return the list of valid instance statuses supported by this TypeDef.
     *
     * @return InstanceStatus array of supported status values.
     */
    public List<InstanceStatus> getValidInstanceStatusList()
    {
        if (validInstanceStatusList == null)
        {
            return null;
        }
        else
        {
            return new ArrayList<>(validInstanceStatusList);
        }
    }


    /**
     * Set up the list of valid instance statuses supported by this TypeDef.
     *
     * @param validInstanceStatusList - InstanceStatus Array
     */
    public void setValidInstanceStatusList(List<InstanceStatus> validInstanceStatusList)
    {
        if (validInstanceStatusList == null)
        {
            this.validInstanceStatusList = null;
        }
        else
        {
            this.validInstanceStatusList = new ArrayList<>(validInstanceStatusList);
        }
    }


    /**
     * Return the initial status setting for an instance of this type.
     *
     * @return InstanceStatus enum
     */
    public InstanceStatus getInitialStatus()
    {
        return initialStatus;
    }


    /**
     * Set up the initial status setting for an instance of this type.
     *
     * @param initialStatus - InstanceStatus enum
     */
    public void setInitialStatus(InstanceStatus initialStatus)
    {
        this.initialStatus = initialStatus;
    }


    /**
     * Return the list of AttributeDefs that define the valid properties for this type of classification.
     *
     * @return AttributeDefs list
     */
    public List<TypeDefAttribute> getPropertiesDefinition()
    {
        if (propertiesDefinition == null)
        {
            return null;
        }
        else
        {
            return new ArrayList<>(propertiesDefinition);
        }
    }


    /**
     * Set up the list of AttributeDefs that define the valid properties for this type of classification.
     *
     * @param propertiesDefinition - AttributeDefs list
     */
    public void setPropertiesDefinition(List<TypeDefAttribute> propertiesDefinition)
    {
        if (propertiesDefinition == null)
        {
            this.propertiesDefinition = null;
        }
        else
        {
            this.propertiesDefinition = new ArrayList<>(propertiesDefinition);
        }
    }


    /**
     * Standard toString method.
     *
     * @return JSON style description of variables.
     */
    @Override
    public String toString()
    {
        return "TypeDef{" +
                "superType=" + superType +
                ", description='" + description + '\'' +
                ", descriptionGUID='" + descriptionGUID + '\'' +
                ", origin='" + origin + '\'' +
                ", createdBy='" + createdBy + '\'' +
                ", updatedBy='" + updatedBy + '\'' +
                ", createTime=" + createTime +
                ", updateTime=" + updateTime +
                ", options=" + options +
                ", externalStandardMappings=" + externalStandardMappings +
                ", validInstanceStatusList=" + validInstanceStatusList +
                ", initialStatus=" + initialStatus +
                ", propertiesDefinition=" + propertiesDefinition +
                ", version=" + version +
                ", versionName='" + versionName + '\'' +
                ", category=" + category +
                ", guid='" + guid + '\'' +
                ", name='" + name + '\'' +
                '}';
    }
}

File: intg/src/main/python/apache_atlas/model/typedef.py
#!/usr/bin/env/python

#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import enum


class AtlasTypesDef:

    def __init__(self, enumDefs=None, structDefs=None, classificationDefs=None,
                 entityDefs=None, relationshipDefs=None, businessMetadataDefs=None):
        self.enumDefs             = enumDefs if enumDefs is not None else []
        self.structDefs           = structDefs if structDefs is not None else []
        self.classificationDefs   = classificationDefs if classificationDefs is not None else []
        self.entityDefs           = entityDefs if entityDefs is not None else []
        self.relationshipDefs     = relationshipDefs if relationshipDefs is not None else []
        self.businessMetadataDefs = businessMetadataDefs if businessMetadataDefs is not None else []


class AtlasBaseTypeDef:

    def __init__(self, category=None, guid=None, createdBy=None, updatedBy=None, createTime=None, updateTime=None,
                 version=None, name=None, description=None, typeVersion=None, serviceType=None, options=None):
        self.category    = category
        self.guid        = guid
        self.createdBy   = createdBy
        self.updatedBy   = updatedBy
        self.createTime  = createTime
        self.updateTime  = updateTime
        self.version     = version
        self.name        = name
        self.description = description
        self.typeVersion = typeVersion
        self.serviceType = serviceType
        self.options     = options if options is not None else {}


category_enum = enum.Enum('category_enum', 'PRIMITIVE OBJECT_ID_TYPE ENUM STRUCT CLASSIFICATION ENTITY ARRAY MAP RELATIONSHIP BUSINESS_METADATA', module=__name__)


class AtlasEnumDef(AtlasBaseTypeDef):

    def __init__(self, category=category_enum.ENUM.name, guid=None, createdBy=None, updatedBy=None, createTime=None,
                 updateTime=None, version=None, name=None, description=None, typeVersion=None,
                 serviceType=None, options=None, elementDefs=None, defaultValue=None):

        super().__init__(category, guid, createdBy, updatedBy, createTime, updateTime, version, name, description, typeVersion, serviceType, options)

        self.elementDefs  = elementDefs if elementDefs is not None else []
        self.defaultValue = defaultValue


class AtlasEnumElementDef:

    def __init__(self, value=None, description=None, ordinal=None):
        self.value       = value
        self.description = description
        self.ordinal     = ordinal


class AtlasStructDef(AtlasBaseTypeDef):

    def __init__(self, category=category_enum.STRUCT.name, guid=None, createdBy=None, updatedBy=None, createTime=None, updateTime=None, version=None,
                 name=None, description=None, typeVersion=None, serviceType=None, options=None, attributeDefs=None):

        super().__init__(category, guid, createdBy, updatedBy, createTime, updateTime, version, name, description, typeVersion, serviceType, options)

        self.attributeDefs = attributeDefs if attributeDefs is not None else []


class AtlasAttributeDef:
    cardinality_enum = enum.Enum('cardinality_enum', 'SINGLE LIST SET', module=__name__)
    indexType_enum   = enum.Enum('indexType_enum', 'DEFAULT STRING', module=__name__)

    def __init__(self, name=None, typeName=None, isOptional=None, cardinality=None, valuesMinCount=None, valuesMaxCount=None,
                 isUnique=None, isIndexable=None, includeInNotification=None, defaultValue=None, description=None, searchWeight= -1,
                 indexType=None, constraints=None, options=None, displayName=None):

        self.name                  = name
        self.typeName              = typeName
        self.isOptional            = isOptional
        self.cardinality           = cardinality
        self.valuesMinCount        = valuesMinCount
        self.valuesMaxCount        = valuesMaxCount
        self.isUnique              = isUnique
        self.isIndexable           = isIndexable
        self.includeInNotification = includeInNotification
        self.defaultValue          = defaultValue
        self.description           = description
        self.searchWeight          = searchWeight
        self.indexType             = indexType
        self.constraints           = constraints if constraints is not None else []
        self.options               = options if options is not None else {}
        self.displayName           = displayName


class AtlasConstraintDef:

    def __init__(self, type=None, params=None):
        self.type   = type
        self.params = params if params is not None else {}


class AtlasClassificationDef(AtlasStructDef):

    def __init__(self, category=category_enum.CLASSIFICATION.name, guid=None, createdBy=None, updatedBy=None, createTime=None, updateTime=None,
                 version=None, name=None, description=None, typeVersion=None, serviceType=None, options=None,
                 attributeDefs=None, superTypes=None, entityTypes=None, subTypes=None):

        super().__init__(category, guid, createdBy, updatedBy, createTime, updateTime, version,
                         name, description, typeVersion, serviceType, options, attributeDefs)

        self.superTypes  = superTypes if superTypes is not None else set()
        self.entityTypes = entityTypes if entityTypes is not None else set()
        self.subTypes    = subTypes if subTypes is not None else set()


class AtlasEntityDef(AtlasStructDef):
    def __init__(self, category=category_enum.CLASSIFICATION.name, guid=None, createdBy=None, updatedBy=None, createTime=None, updateTime=None,
                 version=None, name=None, description=None, typeVersion=None, serviceType=None, options=None,
                 attributeDefs=None, superTypes=None, subTypes=None, relationshipAttributeDefs=None, businessAttributeDefs=None):

        super().__init__(category, guid, createdBy, updatedBy, createTime, updateTime, version,
                         name, description, typeVersion, serviceType, options, attributeDefs)

        self.superTypes                = superTypes if superTypes is not None else set()
        self.subTypes                  = subTypes if subTypes is not None else set()
        self.relationshipAttributeDefs = relationshipAttributeDefs if relationshipAttributeDefs is not None else []
        self.businessAttributeDefs     = businessAttributeDefs if businessAttributeDefs is not None else {}


class AtlasRelationshipAttributeDef(AtlasAttributeDef):
    cardinality_enum = enum.Enum('cardinality_enum', 'SINGLE LIST SET', module=__name__)
    indexType_enum   = enum.Enum('indexType_enum', 'DEFAULT STRING', module=__name__)

    def __init__(self, name=None, typeName=None, isOptional=None, cardinality=None, valuesMinCount=None,
                 valuesMaxCount=None, isUnique=None, isIndexable=None, includeInNotification=None,
                 defaultValue=None, description=None, searchWeight=None, indexType=None, constraints=None,
                 options=None, displayName=None, relationshipTypeName=None, isLegacyAttribute=None):

        super().__init__(name, typeName, isOptional, cardinality, valuesMinCount, valuesMaxCount, isUnique, isIndexable,
                         includeInNotification, defaultValue, description, searchWeight, indexType, constraints, options, displayName)

        self.relationshipTypeName = relationshipTypeName
        self.isLegacyAttribute    = isLegacyAttribute


class AtlasBusinessMetadataDef(AtlasStructDef):

    def __init__(self, category=category_enum.BUSINESS_METADATA.name, guid=None, createdBy=None, updatedBy=None, createTime=None,
                 updateTime=None, version=None, name=None, description=None, typeVersion=None,
                 serviceType=None, options=None, attributeDefs=None):

        super().__init__(category, guid, createdBy, updatedBy, createTime, updateTime, version,
                         name, description, typeVersion, serviceType, options, attributeDefs)


class AtlasRelationshipDef(AtlasStructDef):
    relationshipCategory_enum = enum.Enum('relationshipCategory_enum', 'ASSOCIATION AGGREGATION COMPOSITION', module=__name__)
    propagateTags_enum        = enum.Enum('propagateTags_enum', 'NONE ONE_TO_TWO TWO_TO_ONE BOTH', module=__name__)

    def __init__(self, category=category_enum.RELATIONSHIP.name, guid=None, createdBy=None, updatedBy=None, createTime=None,
                 updateTime=None, version=None, name=None, description=None, typeVersion=None,
                 serviceType=None, options=None, attributeDefs=None, relationshipCategory=None,
                 relationshipLabel=None, propagateTags=None, endDef1=None, endDef2=None):

        super().__init__(category, guid, createdBy, updatedBy, createTime, updateTime, version,
                         name, description, typeVersion, serviceType, options, attributeDefs)

        self.relationshipCategory = relationshipCategory
        self.relationshipLabel    = relationshipLabel
        self.propagateTags        = propagateTags
        self.endDef1              = endDef1
        self.endDef2              = endDef2


class AtlasRelationshipEndDef:
    cardinality_enum = enum.Enum('cardinality_enum', 'SINGLE LIST SET', module=__name__)

    def __init__(self, type=None, name=None, isContainer=None, cardinality=None, isLegacyAttribute=None, description=None):
        self.type              = type
        self.name              = name
        self.isContainer       = isContainer
        self.cardinality       = cardinality if cardinality is not None else AtlasRelationshipEndDef.cardinality_enum.SINGLE.name
        self.isLegacyAttribute = isLegacyAttribute
        self.description       = description